File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String S;
        while(!(S = br.readLine()).equals(".")) {
            Stack<Character> stack = new Stack<>();
            char[] arr = S.toCharArray();
            boolean isTrue = true;
            for(int i = 0; i < arr.length; i++) {
                if(arr[i] == '(' || arr[i] == '[') stack.add(arr[i]);
                else if(arr[i] == ')') {
                    if(stack.isEmpty() || stack.peek() != '(') {
                        isTrue = false;
                        break;
                    }
                    stack.pop();
                }
                else if(arr[i] == ']') {
                    if(stack.isEmpty() || stack.peek() != '[') {
                        isTrue = false;
                        break;
                    }
                    stack.pop();
                }
            }
            sb.append((isTrue && stack.isEmpty()) ? "yes" : "no").append('\n'); 
        }
        System.out.println(sb.toString().trim());
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js

class Stack {
  constructor() {
    this.arr = [];
  }

  push(item) {
    this.arr.push(item);
  }

  pop() {
    return this.arr.pop();
  }

  peek() {
    return this.arr[this.arr.length - 1];
  }

  getSize() {
    return this.arr.length;
  }

  isEmpty() {
    return this.arr.length === 0;
  }
}

const lines = require('fs').readFileSync('/dev/stdin').toString().split('\n');

const ans = [];

for (let i = 0; i < lines.length; i++) {
  if (lines[i] === '.') break;

  const stack = new Stack();
  let isTrue = true;

  for (let ch of lines[i]) {
    if (ch === '(' || ch === '[') {
      stack.push(ch);
    } else if (ch === ')') {
      if (stack.isEmpty() || stack.peek() !== '(') {
        isTrue = false;
        break;
      }
      stack.pop();
    } else if (ch === ']') {
      if (stack.isEmpty() || stack.peek() !== '[') {
        isTrue = false;
        break;
      }
      stack.pop();
    }
  }

  ans.push((isTrue && stack.isEmpty()) ? 'yes' : 'no');
}
console.log(ans.join('\n'));
